[Merged by Bors] - feat(linear_algebra/finite_dimensional): bases given by finsets

pechersky opened the pull request:
In some cases, it might be more straightforward working with finsets of
the vector space instead of indexed types or carrying around a proof of
set.finite. Also provide a lemma on equal dimension
and basis cardinality lemma that uses a finset basis.

Co-authored-by: Scott Morrison scott.morrison@gmail.com

---
<!-- put comments you want to keep out of the PR commit here -->
Developed based on discussion in
https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/working.20with.20finset.20and.20fintype

pechersky committed:
feat(linear_algebra/finite_dimensional): bases given by finsets

In some cases, it might be more straightforward working with finsets of
the vector space instead of indexed types or carrying around a proof of
set.finite. Also provide a lemma on equal dimension
and basis cardinality lemma that uses a finset basis.

pechersky committed:
feat(linear_algebra/finite_dimension): proof of fin-dim using finset basis

bryangingechen commented on src/linear_algebra/finite_dimensional.lean:
```diff
@@ -111,6 +111,19 @@ begin
+/-- In a finite dimensional space, there exists a finite basis. Provides the basis as a finset.
+This is in contrast to `exists_is_basis_finite`, which provides a set and a `set.finite`.
+-/
+lemma finite_dimensional.exists_is_basis_finset [finite_dimensional K V] :
```
```suggestion
lemma exists_is_basis_finset [finite_dimensional K V] :
```

bryangingechen replied:
Not sure if this is in a commit you forgot to push, but the linter is still complaining about this: https://github.com/leanprover-community/mathlib/runs/765849481#step:5:15

bryangingechen commented on src/linear_algebra/finite_dimensional.lean:
```diff
@@ -196,6 +214,13 @@ begin
+/-- If a vector space is finite-dimensional, then the cardinality of any basis is equal to its
+`findim`. This lemma uses finsets instead of indexed types. -/
+lemma finite_dimensional.findim_eq_card_basis'' [finite_dimensional K V] {b : finset V}
+  (h : is_basis K (subtype.val : (↑b : set V) -> V)) :
```
```suggestion
lemma findim_eq_card_basis'' {b : finset V} (h : is_basis K (subtype.val : (↑b : set V) -> V)) :
```
The docstring doesn't quite match the statement now.

bryangingechen reviewed (commented):
Just some suggestions to silence the linter.

pechersky committed:
chore(linear_algebra/finite_dimensional): remove extra author info

pechersky committed:
chore(linear_algebra/finite_dimensional): remove instance req

Like in `findim_eq_card_basis`, the fin-dim-ness of K V is not necessary
for the proof. The instance requirement is removed, and the docstring is
made to be similar to the docstring for `findim_eq_card_basis`,
highlighting the use of `finset`.

kim-em commented on src/linear_algebra/finite_dimensional.lean:
```diff
@@ -196,6 +214,13 @@ begin
 
+/-- If a vector space has a finite basis, then its dimension is equal to the cardinality of the
+basis. This lemma uses a `finset` instead of indexed types. -/
+lemma finite_dimensional.findim_eq_card_basis'' {b : finset V}
```
```suggestion
lemma finite_dimensional.findim_eq_card_finset_basis {b : finset V}
```
Primes don't help anyone!

kim-em commented on src/linear_algebra/finite_dimensional.lean:
```diff
@@ -133,6 +146,11 @@ lemma of_finite_basis {ι : Type w} [fintype ι] {b : ι → V} (h : is_basis K
 iff_fg.2 $ ⟨finset.univ.image b, by {convert h.2, simp} ⟩
 
+/-- If a vector space has a finite basis, then it is finite-dimensional, finset style. -/
+lemma of_finite_basis' {b : finset V} (h : is_basis K (subtype.val : (↑b : set V) -> V)) :
```
```suggestion
lemma of_finset_basis {b : finset V} (h : is_basis K (subtype.val : (↑b : set V) -> V)) :
```

pechersky committed:
chore(linear_algebra/finite_dimension): relint finset basis lemmas

pechersky committed:
Merge branch 'master' into findim-basis-finset
